build.gradle

```diff
@@ -102,14 +103,61 @@ task configureGroovyAll {
             }
         }
     }
+
+    doLast {
+        def expandedArchives = file("${buildDir}/tmp/expandedArchives")
+        def mergedFolder = new File(expandedArchives, 'merged')
+
+        collectDuplicates(expandedArchives.listFiles())
+            .each { String relative, List<File> files ->
+                def mergedFile = new File(mergedFolder, relative)
+                mergedFile.parentFile.mkdirs()
+                mergedFile.text = merge(relative, files)
+                files*.delete() // TODO doesn't work (manifests are read after this, and even if we delete, they'll still end up in the jar)
+            }
+    }
+}
+
+static collectDuplicates(File[] roots) {
+    Map<String, List<File>> duplicates = [:].withDefault { [] }
+    roots.each { root ->
+        root.eachFileRecurse(FileType.FILES) {
+            def relative = root.toURI().relativize(it.toURI())
+            duplicates[relative.toString()] += it
+        }
+    }
+    return duplicates.findAll { it.value.size() > 1 }
+}
+
+static merge(String relative, List<File> files) {
+    switch (relative) {
+        case ['META-INF/LICENSE', 'META-INF/NOTICE', 'META-INF/services/org.codehaus.groovy.transform.ASTTransformation']:
+            return files*.text.join("\n")
+        case 'META-INF/groovy/org.codehaus.groovy.runtime.ExtensionModule':
+            return """
+                moduleName=groovy-all
+                moduleVersion=${(files[0].text =~ 'moduleVersion=(.+)')[0][1]}
+                extensionClasses=${(files*.text.join('\n') =~ 'extensionClasses=(.+)')*.getAt(1).toSet().join(',')}
+                staticExtensionClasses=${(files*.text.join('\n') =~ 'staticExtensionClasses=(.+)')*.getAt(1).toSet().join(',')}
+            """.stripIndent().trim()
+        case 'META-INF/groovy-release-info.properties':
+            return files[0].text
+
+        case ['META-INF/MANIFEST.MF', 'META-INF/services/org.apache.groovy.plugin.GroovyRunner', 'META-INF/INDEX.LIST']:
+            return '' // TODO should be ignored
```

Comment:
We could use `null` here to distinguish between empty files and files that should be skipped.

Reply:
Yup, that's what I did in the end